File: examples/capture_both.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include "audiotap.h"

static volatile sig_atomic_t g_running = 1;

static void signal_handler(int sig)
{
    (void)sig;
    g_running = 0;
}

static void system_callback(const float *samples, uint32_t frame_count,
                             uint32_t channels, uint64_t host_time, void *userdata)
{
    (void)host_time;
    FILE *fp = (FILE *)userdata;
    fwrite(samples, sizeof(float), frame_count * channels, fp);
}

static void mic_callback(const float *samples, uint32_t frame_count,
                          uint32_t channels, uint64_t host_time, void *userdata)
{
    (void)host_time;
    FILE *fp = (FILE *)userdata;
    fwrite(samples, sizeof(float), frame_count * channels, fp);
}

int main(int argc, char *argv[])
{
    int duration = 10;
    if (argc > 1)
        duration = atoi(argv[1]);
    if (duration <= 0)
        duration = 10;

    float sample_rate = 48000.0f;
    uint32_t sys_channels = 2;
    uint32_t mic_channels = 1;

    FILE *sys_file = fopen("system.pcm", "wb");
    FILE *mic_file = fopen("mic.pcm", "wb");
    if (!sys_file || !mic_file) {
        fprintf(stderr, "Failed to open output files\n");
        return 1;
    }

    audiotap_system_config_t sys_config = {
        .pids = NULL,
        .pid_count = 0,
        .sample_rate = sample_rate,
        .channels = sys_channels,
        .mute = 0,
        .callback = system_callback,
        .userdata = sys_file,
    };

    audiotap_mic_config_t mic_config = {
        .sample_rate = sample_rate,
        .channels = mic_channels,
        .callback = mic_callback,
        .userdata = mic_file,
    };

    printf("Requesting microphone permission...\n");
    audiotap_permission_t perm = audiotap_request_mic_permission();
    if (perm != AUDIOTAP_PERMISSION_GRANTED) {
        fprintf(stderr, "Microphone permission not granted (status=%d).\n"
                        "Grant access in System Settings > Privacy & Security > Microphone.\n",
                (int)perm);
    }

    printf("Creating system audio tap...\n");
    audiotap_t *sys_tap = audiotap_create_system(&sys_config);
    if (!sys_tap) {
        fprintf(stderr, "Failed to create system tap (need macOS 14.2+ and audio permissions)\n");
    }

    printf("Creating microphone tap...\n");
    audiotap_t *mic_tap = audiotap_create_mic(&mic_config);
    if (!mic_tap) {
        fprintf(stderr, "Failed to create mic tap (no input device?)\n");
    }

    if (!sys_tap && !mic_tap) {
        fprintf(stderr, "No audio sources available. Exiting.\n");
        fclose(sys_file);
        fclose(mic_file);
        return 1;
    }

    if (sys_tap) {
        int err = audiotap_start(sys_tap);
        if (err)
            fprintf(stderr, "System tap start error: %s (%d)\n", audiotap_error_string(err), err);
        else
            printf("System audio capture started.\n");
    }

    if (mic_tap) {
        int err = audiotap_start(mic_tap);
        if (err)
            fprintf(stderr, "Mic tap start error: %s (%d)\n", audiotap_error_string(err), err);
        else
            printf("Microphone capture started.\n");
    }

    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    printf("Recording for %d seconds (Ctrl+C to stop early)...\n", duration);
    for (int i = 0; i < duration && g_running; i++)
        sleep(1);

    printf("\nStopping...\n");

    if (sys_tap)
        audiotap_destroy(sys_tap);
    if (mic_tap)
        audiotap_destroy(mic_tap);

    fclose(sys_file);
    fclose(mic_file);

    printf("Done. Playback with:\n");
    printf("  ffplay -f f32le -ar %.0f -ac %u system.pcm\n", sample_rate, sys_channels);
    printf("  ffplay -f f32le -ar %.0f -ac %u mic.pcm\n", sample_rate, mic_channels);

    return 0;
}

File: python/examples/capture_both.py
#!/usr/bin/env python3
"""Capture system audio and microphone simultaneously — Python port of capture_both.c.

Writes raw float32 PCM to system.pcm and mic.pcm.

Usage:
    uv run examples/capture_both.py [duration_seconds]
"""

from __future__ import annotations

import signal
import sys
import time

import audiotap

SAMPLE_RATE = 48000.0
SYSTEM_CHANNELS = 2
MIC_CHANNELS = 1


def main() -> int:
    duration = int(sys.argv[1]) if len(sys.argv) > 1 else 10
    if duration <= 0:
        duration = 10

    sys_file = open("system.pcm", "wb")
    mic_file = open("mic.pcm", "wb")

    def system_callback(samples: bytes, frame_count: int, channels: int, host_time: int) -> None:
        sys_file.write(samples)

    def mic_callback(samples: bytes, frame_count: int, channels: int, host_time: int) -> None:
        mic_file.write(samples)

    # Request mic permission (blocks for user dialog)
    print("Requesting microphone permission...")
    perm = audiotap.request_mic_permission()
    if perm != audiotap.Permission.GRANTED:
        print(
            f"Microphone permission not granted (status={perm.name}).\n"
            "Grant access in System Settings > Privacy & Security > Microphone.",
            file=sys.stderr,
        )

    # Create taps
    sys_tap = None
    mic_tap = None

    print("Creating system audio tap...")
    try:
        sys_tap = audiotap.SystemTap(
            callback=system_callback,
            sample_rate=SAMPLE_RATE,
            channels=SYSTEM_CHANNELS,
        )
    except audiotap.AudioTapError as e:
        print(f"Failed to create system tap: {e}", file=sys.stderr)

    print("Creating microphone tap...")
    try:
        mic_tap = audiotap.MicTap(
            callback=mic_callback,
            sample_rate=SAMPLE_RATE,
            channels=MIC_CHANNELS,
        )
    except audiotap.AudioTapError as e:
        print(f"Failed to create mic tap: {e}", file=sys.stderr)

    if sys_tap is None and mic_tap is None:
        print("No audio sources available. Exiting.", file=sys.stderr)
        sys_file.close()
        mic_file.close()
        return 1

    # Start capturing
    if sys_tap is not None:
        try:
            sys_tap.start()
            print("System audio capture started.")
        except audiotap.AudioTapError as e:
            print(f"System tap start error: {e}", file=sys.stderr)

    if mic_tap is not None:
        try:
            mic_tap.start()
            print("Microphone capture started.")
        except audiotap.AudioTapError as e:
            print(f"Mic tap start error: {e}", file=sys.stderr)

    # Wait for duration or Ctrl+C
    running = True

    def stop_handler(sig, frame):
        nonlocal running
        running = False

    signal.signal(signal.SIGINT, stop_handler)
    signal.signal(signal.SIGTERM, stop_handler)

    print(f"Recording for {duration} seconds (Ctrl+C to stop early)...")
    elapsed = 0
    while elapsed < duration and running:
        time.sleep(1)
        elapsed += 1

    print("\nStopping...")

    if sys_tap is not None:
        sys_tap.destroy()
    if mic_tap is not None:
        mic_tap.destroy()

    sys_file.close()
    mic_file.close()

    print("Done. Playback with:")
    print(f"  ffplay -f f32le -ar {SAMPLE_RATE:.0f} -ac {SYSTEM_CHANNELS} system.pcm")
    print(f"  ffplay -f f32le -ar {SAMPLE_RATE:.0f} -ac {MIC_CHANNELS} mic.pcm")

    return 0


if __name__ == "__main__":
    sys.exit(main())
